src/ratelimit/RateLimitController.ts

```diff
@@ -0,0 +1,194 @@
+import type { Patch } from 'immer';
+
+import { BaseController } from '../BaseControllerV2';
+
+import type { RestrictedControllerMessenger } from '../ControllerMessenger';
+
+/**
+ * @type RateLimitState
+ * @property requests - Object containing number of requests in a given interval for each origin and api type combination
+ */
+export type RateLimitState<ApiType extends string> = {
+  requests: Record<ApiType, Record<string, number>>;
+};
+
+export interface CallArgs<ApiType> {
+  /**
+   * Enum type to determine API type.
+   */
+  type: ApiType;
+
+  /**
+   * Args passed to the API call
+   */
+  args?: unknown[];
+}
+
+export type RateLimitWrapper =
+  | { isRateLimited: false; result: unknown }
+  | { isRateLimited: true };
+
+const name = 'RateLimitController';
+
+export type RateLimitStateChange<ApiType extends string> = {
+  type: `${typeof name}:stateChange`;
+  payload: [RateLimitState<ApiType>, Patch[]];
+};
+
+export type GetRateLimitState<ApiType extends string> = {
+  type: `${typeof name}:getState`;
+  handler: () => RateLimitState<ApiType>;
+};
+
+export type CallApi<ApiType extends string> = {
+  type: `${typeof name}:call`;
+  handler: RateLimitController<ApiType>['call'];
+};
+
+export type ControllerActions<ApiType extends string> =
+  | GetRateLimitState<ApiType>
+  | CallApi<ApiType>;
+
+export type RateLimitMessenger<
+  ApiType extends string
+> = RestrictedControllerMessenger<
+  typeof name,
+  ControllerActions<ApiType>,
+  RateLimitStateChange<ApiType>,
+  never,
+  never
+>;
+
+const metadata = {
+  requests: { persist: false, anonymous: false },
+};
+
+/**
+ * Controller with logic for rate-limiting API endpoints per requesting origin.
+ */
+export class RateLimitController<ApiType extends string> extends BaseController<
+  typeof name,
+  RateLimitState<ApiType>,
+  RateLimitMessenger<ApiType>
+> {
+  private implementations;
+
+  private rateLimitTimeout;
+
+  private rateLimitCount;
+
+  /**
+   * Creates a RateLimitController instance.
+   *
+   * @param options - Constructor options.
+   * @param options.messenger - A reference to the messaging system.
+   * @param options.state - Initial state to set on this controller.
+   * @param options.implementations - Mapping from API type to API implementation.
+   * @param options.rateLimitTimeout - The time window in which the rate limit is applied.
+   * @param options.rateLimitCount - The amount of notifications an origin can show in the rate limit time window.
+   */
+  constructor({
+    // @todo Pick some sane defaults for this
+    rateLimitTimeout = 5000,
+    rateLimitCount = 3,
+    messenger,
+    state,
+    implementations,
+  }: {
+    rateLimitTimeout?: number;
+    rateLimitCount?: number;
+    messenger: RateLimitMessenger<ApiType>;
+    state?: Partial<RateLimitState<ApiType>>;
+    implementations: Record<ApiType, (...args: unknown[]) => unknown>;
```

Comment:
This doesn't seem practical to use. In effect these implementations can't be used by the controller unless they happen to accept `unknown` or `any` as parameters. Try updating the unit tests to pass in a function with specific parameter types to see what I mean. Effectively this erases the types of the underlying method, so using would require lots of inconvenient casting and the loss of type safety.

When I suggested making this controller generic, what I had in mind was making it generic over the underlying implementation so their types would be preserved. We need to preserve the types for this to be practical.

I made an example of this here: https://github.com/MetaMask/controllers/tree/rate-limit-controller-types-example
Not sure if it works, and I didn't update the tests, so my apologies if it's broken. But that's generally what I was thinking originally.

Reply:
@Gudahtt Hmm, this is a good idea, I think. I can explore this further tomorrow! Thanks!